webapp/commonComponents/expression/expressionVariables.js

```diff
@@ -41,37 +41,37 @@ const getSqlVariables = (nodeDef, lang) => {
   }))
 }
 
-const getChildDefVariables = (survey, nodeDefContext, mode, lang) => {
+const getChildDefVariables = (survey, nodeDefContext, nodeDefCurrent, mode, lang) => R.pipe(
+  Survey.getNodeDefChildren(nodeDefContext),
+  //exclude nodes that reference the current one
+  R.unless(
+    R.always(R.isNil(nodeDefCurrent)),
+    R.reject(nodeDef => Survey.isNodeDefDependentOn(NodeDef.getUuid(nodeDef), NodeDef.getUuid(nodeDefCurrent))(survey)),
+  ),
+  R.map(childDef => {
+    if (!Expression.isValidExpressionType(childDef))
+      return null
+    else if (mode === Expression.modes.sql)
+      return getSqlVariables(childDef, lang)
+    else if (mode === Expression.modes.json)
+      return getJsVariables(childDef, lang)
+  }),
+  R.flatten,
+  R.reject(R.isNil),
+)(survey)
 
-  return R.pipe(
-    Survey.getNodeDefChildren(nodeDefContext),
-    R.map(childDef => {
-      if (!Expression.isValidExpressionType(childDef))
-        return null
-      else if (mode === Expression.modes.sql)
-        return getSqlVariables(childDef, lang)
-      else if (mode === Expression.modes.json)
-        return getJsVariables(childDef, lang)
-    }),
-    R.flatten,
-    R.reject(R.isNil),
+export const getVariables = (survey, nodeDefContext, nodeDefCurrent, mode, preferredLang) => {
+  const lang = Survey.getLanguage(preferredLang)(Survey.getSurveyInfo(survey))
+  const surveyWithDependencies = R.pipe(
+    Survey.buildDependencyGraph,
+    graph => Survey.assocDependencyGraph(graph)(survey)
```

Comment:
you are building the dependency graph every time an expression is edited. is that what we really want? can this cause performance issues?

Reply:
The dependency graph can change whenever a nodeDef expression is changed; we can generate it one time and associate it to the survey in the state, but we need also to update it when needed. How can we do it?

Reply:
what do you mean how we can do it? we implement it